### ISSUE TITLE
compile against opencascade 7.8.0 fails

### ISSUE THREAD

**Issue by adrianinsaval** [IfcOpenShell/IfcOpenShell#4393]
I get this error when trying to compile against occt 7.8:
```
ninja: error: '/app/lib/libTKSTEP.so', needed by 'ifcwrap/_ifcopenshell_wrapper.so', missing and no known rule to make it
Error: module IfcOpenShell: Child process exited with code 1

```

this is because some occt libs have been reorginized, see https://dev.opencascade.org/doc/overview/html/occt__upgrade.html#upgrade_occt780

cc @bgbsww while this is not a hard dependency for freecad main we ship it in our bundles, I think it's required by some arch wb functionalities and/or some of the addons that will be combined into it (@yorikvanhavre would know better what it's used for). If you have some spare time it would be nice to get help in migrating this to occt 7.8 too

**Comment by bgbsww** [comment]
Hmm.  I bet the issue is https://github.com/FreeCAD/FreeCAD/blob/e0ecd213b2c34673cb58b9ee04f3a3db152e4b09/src/Mod/TechDraw/App/Makefile.am#L45  which doesn't use the libraries that were located at https://github.com/FreeCAD/FreeCAD/blob/e0ecd213b2c34673cb58b9ee04f3a3db152e4b09/cMake/FindOCC.cmake#L159-L164  I don't understand at a glance why there's a standalone Makefile.am here or what the correct path is - presumably update this to cmake?  Looks like an anomoly.

**Comment by aothms** [comment]
We have a hardcoded list here. https://github.com/IfcOpenShell/IfcOpenShell/blob/bce0f0d8d4d06015bbc82e9de71799fd3d02c676/cmake/CMakeLists.txt#L379-L385

Some help would be very welcome if you know already what to do for 7.8 compatibility.

**Comment by bgbsww** [comment]
Yes, see the second link above.  You need to pull the TKSTEP and TKSTEPBase out of the OPENCASCADE_LIBRARY_NAMES and then add a following if/else/endif clause that inserts either those two,  or else TKDESTEP depending on the OCC_VERSION_STRING.

**Comment by adrianinsaval** [comment]
looks like it works, although I did had to add the logic to define OCC_VERSION_STRING, copied straight from freecad's FIndOCC

**Comment by bgbsww** [comment]
Oh, yeah.  Duh.  CMake isn't going to magically define OCC package strings.  That's the right approach, glad you got it all going.

**Comment by yorikvanhavre** [comment]
BTW - given the rapid evolution of IfcOpenShell and thanks to the build service it has, I'm working on a tool to allow users to download and update their version of IfcOpenShell from inside FreeCAD. We could then stop bundling it with FreeCAD.

@adrianinsaval to reply your question above, IfcOpenShell is used 1) to import/export IFC files in Arch/BIM WB, and 2) as an engine when using NativeIFC. It is never a hard requirement to use Arch/BIM, only if you want to work with IFC.

The fact that it is bundled in our installer comes from a time when it was somewhat harder to get your hands on a compatible build of IfcOpenShell, and many windows users had asked for it... And we can't say no to users... :sweat_smile:

**Comment by adrianinsaval** [comment]
> The fact that it is bundled in our installer comes from a time when it was somewhat harder to get your hands on a compatible build of IfcOpenShell, and many windows users had asked for it... And we can't say no to users... 😅

don't we need to always match the occt and python versions?

**Comment by aothms** [comment]
The typical way shape data is exchanged between ifcopenshell and other libs is through string serializations of the occt topology. So there isn't a strict requirement on matching occt versions. Somewhere recently (7.7? / 7.8?) a new version v3 of the occt serialization format is introduced though, so that might result in incompatibilities if one keeps using the statically linked ifcopenshell binaries with 7.5. But the scenarios of passing topology data from freecad to ifcopenshell are probably limited.

**Comment by yorikvanhavre** [comment]
> But the scenarios of passing topology data from freecad to ifcopenshell are probably limited.

Indeed, that does not happen. For that, we use the API to create IFC objects. Serialization to strings is only used for ifcopenshell -> freecad.

**Comment by martin-bernhard** [comment]
> The typical way shape data is exchanged between ifcopenshell and other libs is through string serializations of the occt topology. So there isn't a strict requirement on matching occt versions. Somewhere recently (7.7? / 7.8?) a new version v3 of the occt serialization format is introduced though, so that might result in incompatibilities if one keeps using the statically linked ifcopenshell binaries with 7.5. But the scenarios of passing topology data from freecad to ifcopenshell are probably limited.

To avoid incompatibility between serialized shapes from OCCT, you can explicitly use the v1 format, even with OCCT 7.8, so that it can be read with older versions of OCCT, too.

**Comment by aothms** [comment]
> To avoid incompatibility between serialized shapes from OCCT, you can explicitly use the v1 format, even with OCCT 7.8, so that it can be read with older versions of OCCT, too.

True, but that requires a change in the IfcOpenShell C++ code, which is troublesome for most users. 

@martin-bernhard if you have the code ready a PR would be appreciated :) https://github.com/IfcOpenShell/IfcOpenShell/blob/453ef437d3656859064eb30ee9de16f1d70bd8f8/src/ifcgeom/kernels/opencascade/OpenCascadeConversionResult.cpp#L245

**Comment by martin-bernhard** [comment]
I dont't have a running version, but you should mainly replace line 245 by the following line:
```cpp
BRepTools::Write(s, sstream, true, false, TopTools_FormatVersion_VERSION_1);
```
The constant `TopTools_FormatVersion_VERSION_1` is defined in the `TopTools_FormatVersion.hxx` header file with the integer value 1.
